src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +22,206 @@ namespace System
+            => TrimHelper(ref span, ref trimElements, false, true);
+
+        /// <summary>
+        /// Conditionally removes all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <param name="trimStart">If set to True, trims the start.</param>
+        /// <param name="trimEnd">If set to True, trims the end.</param>
+        private static ReadOnlySpan<T> TrimHelper<T>(ref ReadOnlySpan<T> span, ref ReadOnlySpan<T> trimElements, bool trimStart, bool trimEnd)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty || trimElements.IsEmpty)
+                return span;
+
+            int start = 0;
+            int end = span.Length - 1;
+
+            if (trimStart)
+            {
+                for (; start < span.Length; start++)
+                {
+                    for (int i = 0; i < trimElements.Length; i++)
+                    {
+                        if (trimElements[i] == null)
+                        {
+                            if (span[start] != null)
+                                goto Next;
+                        }
+                        else
+                        {
+                            if (span[start] != null && !span[start].Equals(trimElements[i]))
+                                goto Next;
+                        }
+                    }
+                    break;
+                Next:
+                    ;
+                }
+            }
+
+            if (trimEnd)
+            {
+                // Stop searching once we reach start
+                for (; end >= start; end--)
+                {
+                    for (int i = 0; i < trimElements.Length; i++)
+                    {
+                        if (trimElements[i] == null)
+                        {
+                            if (span[end] != null)
+                                goto Next;
+                        }
+                        else
+                        {
+                            if (span[end] != null && !span[end].Equals(trimElements[i]))
+                                goto Next;
+                        }
+                    }
+                    break;
+                Next:
+                    ;
+                }
+            }
+
+            return span.Slice(start, end - start + 1);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified byte.
+        /// </summary>
+        /// <param name="span">The source span from which the byte is removed.</param>
+        /// <param name="trimValue">The specified byte to look for and remove.</param>
+        public static ReadOnlySpan<byte> TrimStart(this ReadOnlySpan<byte> span, byte trimValue)
```

Comment:
As an aside, should we not be adding closed overloads for `byte`, like we do for `char` ?
Else we'll needlessly incur the `null` checks, etc shown in the previous code

Reply:
> Else we'll needlessly incur the `null` checks, etc shown in the previous code

Those should be optimized out.

Reply:
Reverted